### PR TITLE
Add regexp template grammar

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

Fixes: [#lang-39956](https://github.com/ballerina-platform/ballerina-lang/issues/39956)
Fixes: [#lang-39949](https://github.com/ballerina-platform/ballerina-lang/issues/39949)

## Goals

- This will color the syntaxes inside the regexp template.
- Mark/highlight the bracket bounds

## Samples
Screenshot 2023-04-20 at 10 46 52

### CODE REVIEW

**Comment by LakshanWeerasinghe** [comment]
SS after running `npm run test` command. 
Screenshot 2023-04-20 at 11 02 30

**Comment by LakshanWeerasinghe** [comment]
> SS after running `npm run test` command. Screenshot 2023-04-20 at 11 02 30

This failure happen because of a change I made to the `stringTemplate` grammar. I have reverted that and force pushed the changes.
Screenshot 2023-04-20 at 11 13 43

**Comment by LakshanWeerasinghe** [comment]
> SS after running `npm run test` command. Screenshot 2023-04-20 at 11 02 30

This failure happen because of a change I made to the `stringTemplate` grammar. I have reverted that and force pushed the changes.
Screenshot 2023-04-20 at 11 13 43

File: syntaxes/ballerina.YAML-tmLanguage [hunk 1]
```diff
@@ -1712,8 +1723,132 @@ repository:
...
+      match: '[?+*]|\{(\d+,\d+|\d+,|,\d+|\d+)\}\??'
+    - name: keyword.operator.or.regexp.ballerina
+      match: \|
+    - name: meta.group.assertion.regexp.ballerina
+      begin: (\()
+      beginCaptures:
+        '1': { name: punctuation.definition.group.regexp.ballerina }
+      end: (\))
+      endCaptures:
+        '1': {name: punctuation.definition.group.regexp.ballerina }
```
Comment (SasinduDilshara) [inline syntaxes/ballerina.YAML-tmLanguage:1761]:
```suggestion
        '1': { name: punctuation.definition.group.regexp.ballerina }
```

File: syntaxes/ballerina.YAML-tmLanguage [hunk 2]
```diff
@@ -1712,8 +1723,132 @@ repository:
...
+      - include: '#flags-on-off'
+      - include: '#unicode-property-escape'
+    - name: constant.other.character-class.set.regexp.ballerina
+      begin: (\[)(\^)?
+      beginCaptures:
+        '1': { name: punctuation.definition.character-class.start.regexp.ballerina }
+        '2': { name: keyword.operator.negation.regexp.ballerina }
+      end: (\])
+      endCaptures:
+        '1': {name: punctuation.definition.character-class.end.regexp.ballerina }
```
Comment (SasinduDilshara) [inline syntaxes/ballerina.YAML-tmLanguage:1773]:
```suggestion
        '1': { name: punctuation.definition.character-class.end.regexp.ballerina }
```

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

File: syntaxes/ballerina.YAML-tmLanguage [hunk 3]
```diff
@@ -1712,8 +1723,132 @@ repository:
...
+      beginCaptures:
+        '1': { name: punctuation.other.non-capturing-group-begin.regexp.ballerina }
+        '2': { name: keyword.other.non-capturing-group.flags-on.regexp.ballerina }
+        '3': { name: punctuation.other.non-capturing-group.off.regexp.ballerina }
+        '4': { name: keyword.other.non-capturing-group.flags-off.regexp.ballerina }
+        '5': { name: punctuation.other.non-capturing-group-end.regexp.ballerina }
+      end: ()
+      name: constant.other.flag.regexp.ballerina
+      patterns:
+      - include: '#regexp'
```
Comment (SasinduDilshara) [inline syntaxes/ballerina.YAML-tmLanguage:1854]: Need a new line at the end of the file.

File: test/resources/config/regexp.bal [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    string pattern = "ABC";
+    
+    string:RegExp _ = re `abc`;
+    string:RegExp _ = re `(abc)`;
+    string:RegExp _ = re `[abc]`;
+
+    string:RegExp _ = re `\p{gc=Ll}|\P{Lm}|\p{sc=Latin}`;
+    string:RegExp _ = re `\u{12FFFF}`;
+
+    string:RegExp _ = re `${pattern}`;
```
Comment (SasinduDilshara) [inline test/resources/config/regexp.bal:12]: In here, the colours of the `$` and the Curley brackets are different.
Don't we need to give the same colours for `$` and `{ }`, like we do in the `\p{}`.

File: test/resources/config/regexp.bal [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    string:RegExp _ = re `(?-:abc)`;
+    string:RegExp _ = re `(?i-:abc)`;
+    string:RegExp _ = re `(?im-sx:abc)`;
+
+    string:RegExp _ = re `(abc)|${pattern}|[abc]|(?i:abc)`;
+    string:RegExp _ = re `[((abc))))\]]`;
+
+    string:RegExp _ = re `[abc\w\p{Ll}\u{FF}]`;
+
+}
```
Comment (SasinduDilshara) [inline test/resources/config/regexp.bal:24]: Need a new line at the end of the file.

**Review by SasinduDilshara — CHANGES_REQUESTED** [review]

**Comment by gigara** [comment]
>  Screenshot 2023-04-20 at 11 13 43

@LakshanWeerasinghe Why is the new test case is missing?

File: test/resources/config/regexp.bal (hunk 4 above)
Comment (LakshanWeerasinghe) [reply test/resources/config/regexp.bal:12]: no. Those two comes under two different scopes and they have different meanings.

`\p{Ll}` and `\u{FFF}` cases should give the same color.

File: syntaxes/ballerina.tmLanguage [hunk 6]
```diff
@@ -5234,6 +5282,404 @@
...
+                <key>include</key>
+                <string>#regexp</string>
+              </dict>
+            </array>
+          </dict>
+        </array>
+      </dict>
     </dict>
   </dict>
 </plist>
```
Comment (sanjana) [inline syntaxes/ballerina.tmLanguage:5685]: Missing newline at the end of the file.

File: test/resources/snapshots/regexp.bal.snap [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+#                                           ^ source.ballerina meta.function.ballerina meta.block.ballerina meta.var.expr.ballerina regexp.template.ballerina constant.other.character-class.set.regexp.ballerina keyword.other.unicode-value.ballerina punctuation.other.unicode-value.end.regexp.ballerina
+#                                            ^ source.ballerina meta.function.ballerina meta.block.ballerina meta.var.expr.ballerina regexp.template.ballerina constant.other.character-class.set.regexp.ballerina punctuation.definition.character-class.end.regexp.ballerina
+#                                             ^ source.ballerina meta.function.ballerina meta.block.ballerina meta.var.expr.ballerina regexp.template.ballerina
+#                                              ^ source.ballerina meta.function.ballerina meta.block.ballerina punctuation.terminator.statement.ballerina
+>
+>}
+#^ source.ballerina meta.function.ballerina meta.block.ballerina punctuation.definition.block.ballerina
```
Comment (sanjana) [inline test/resources/snapshots/regexp.bal.snap:311]: Missing newline at the EOF.

**Review by sanjana — COMMENTED** [review]
Formatting issues.

**Comment by LakshanWeerasinghe** [comment]
> > Screenshot 2023-04-20 at 11 13 43
> 
> @LakshanWeerasinghe Why is the new test case is missing?

Updated.

Screenshot 2023-04-25 at 09 35 22

File: syntaxes/ballerina.YAML-tmLanguage [hunk 8]
```diff
@@ -1712,8 +1723,133 @@ repository:
...
+
```
Comment (SasinduDilshara) [inline syntaxes/ballerina.YAML-tmLanguage:1855]: Missing the newline in here

File: test/unit-test.ts [hunk 9]
```diff
@@ -202,4 +202,9 @@ describe('Unit tests', () => {
         return runTextMateTest('json');
     });
 
+    it('regexp snapshot test', () => {
+        return runTextMateTest('regexp');
+    });
+
+
```
Comment (sanjana) [inline test/unit-test.ts:209]: Superfluous newline.
```suggestion
```

File: syntaxes/ballerina.tmLanguage (hunk 6 above)
Comment (SasinduDilshara) [inline syntaxes/ballerina.tmLanguage:5685]: Missing the newline here

File: test/resources/config/regexp.bal [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+    string:RegExp _ = re `(?-:abc)`;
+    string:RegExp _ = re `(?i-:abc)`;
+    string:RegExp _ = re `(?im-sx:abc)`;
+
+    string:RegExp _ = re `(abc)|${pattern}|[abc]|(?i:abc)`;
+    string:RegExp _ = re `[((abc))))\]]`;
+
+    string:RegExp _ = re `[abc\w\p{Ll}\u{FF}]`;
+
```
Comment (SasinduDilshara) [inline test/resources/config/regexp.bal:23]: We can remove this new line

File: test/unit-test.ts (hunk 9 above)
Comment (SasinduDilshara) [inline test/unit-test.ts:209]: We can remove these two new lines

File: test/resources/config/regexp.bal (hunk 10 above)
Comment (sanjana) [inline test/resources/config/regexp.bal:23]: Superfluous newline.
```suggestion
```

File: syntaxes/ballerina.YAML-tmLanguage (hunk 8 above)
Comment (sanjana) [inline syntaxes/ballerina.YAML-tmLanguage:1855]: Check why the ending newline is not detected here.

File: test/resources/config/regexp.bal (hunk 10 above)
Comment (SasinduDilshara) [inline test/resources/config/regexp.bal:23]: Shall we add a one more example with character ranges.
We need tests for the module level regex definitions as well.

File: syntaxes/ballerina.YAML-tmLanguage [hunk 11]
```diff
@@ -1712,8 +1723,133 @@ repository:
     - include: '#maps'
     - include: '#paranthesised'
     - include: '#paranthesisedBracket'
+    - include: '#regex'
+
```
Comment (sanjana) [inline syntaxes/ballerina.YAML-tmLanguage:1727]: Is this an unnecessary newline?

File: test/resources/config/regexp.bal [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    string:RegExp _ = re `${pattern}`;
+    string:RegExp _ = re `\w\s\n`;
+
+    string:RegExp _ = re `(?-:abc)`;
+    string:RegExp _ = re `(?i-:abc)`;
+    string:RegExp _ = re `(?im-sx:abc)`;
+
+    string:RegExp _ = re `(abc)|${pattern}|[abc]|(?i:abc)`;
+    string:RegExp _ = re `[((abc))))\]]`;
```
Comment (sanjana) [inline test/resources/config/regexp.bal:20]: Better separate the negative syntax test cases. Can use a comment within this bal file itself. 
Also, check how this works for multi-lines, where the regexp is spanning across multiple lines.

File: test/resources/config/regexp.bal (hunk 4 above)
Comment (sanjana) [inline test/resources/config/regexp.bal:12]: Another possible test case is to use a function call within the interpolation. Can verify the function call expression's brackets - () are appropriately highlighted.

E.g.:
```
function myFn() returns string {
   return "abc";
}
```

```
string:RegExp _ = re `${myFn()}`;
```

File: test/resources/config/regexp.bal (hunk 10 above)
Comment (sanjana) [inline test/resources/config/regexp.bal:23]: Another possible test case is to use a record/object field access within the interpolation and verify the syntax highlighting.

E.g.:

```
type MyRecord record {
   string myStr = "aaa";
};

MyRecord rA;
```

```
string:RegExp _ = re `${rA.myStr}`;
```

Can apply this for functions defined with objects as well.

File: test/resources/config/regexp.bal (hunk 12 above)
Comment (sanjana) [reply test/resources/config/regexp.bal:20]: E.g.:

```
string:RegExp _ = re `This \n
is a \r\n
multi-line
comment
`;
```

File: syntaxes/ballerina.tmLanguage (hunk 6 above)
Comment (LakshanWeerasinghe) [reply syntaxes/ballerina.tmLanguage:5685]: This file is generated one.

File: syntaxes/ballerina.YAML-tmLanguage (hunk 11 above)
Comment (LakshanWeerasinghe) [reply syntaxes/ballerina.YAML-tmLanguage:1727]: yes

File: test/resources/config/regexp.bal (hunk 12 above)
Comment (LakshanWeerasinghe) [reply test/resources/config/regexp.bal:20]: We are not adding negative tests.

**Comment by LakshanWeerasinghe** [comment]
Final Output:
Screenshot 2023-05-11 at 10 46 33

Test Results:
Screenshot 2023-05-11 at 10 52 20

File: test/resources/config/regexp.bal (hunk 12 above)
Comment (LakshanWeerasinghe) [reply test/resources/config/regexp.bal:29]: Screenshot 2023-05-11 at 10 55 52

It's work for multi lines.

**Comment by SasinduDilshara** [comment]
> Final Output: Screenshot 2023-05-11 at 10 46 33
> 
> Test Results: Screenshot 2023-05-11 at 10 52 20

String inclusion contains same color for `$` and `{}`.
We need to do the same for `${}`

**Comment by LakshanWeerasinghe** [comment]
Test Result:
Screenshot 2023-05-14 at 22 26 11

Current Syntax coloring after interpolation coloring:
Screenshot 2023-05-14 at 22 26 28

**Comment by SasinduDilshara** [comment]
> Test Result: Screenshot 2023-05-14 at 22 26 11
> 
> Current Syntax coloring after interpolation coloring: Screenshot 2023-05-14 at 22 26 28

LGTM

File: syntaxes/ballerina.tmLanguage (hunk 6 above)
Comment (SasinduDilshara) [reply syntaxes/ballerina.tmLanguage:5701]: There should be a new line in here. Can you created a issue for this. A new line should be at the end while generating

**Review by SasinduDilshara — APPROVED** [review]

**Pull request merged**